test/Microsoft.OData.ModelBuilder.Tests/Commons/MockType.cs

```diff
@@ -85,8 +85,46 @@ public MockType Property(Type propertyType, string propertyName, params Attribut
+            // So we have to mock the GetCustomAttributes methods. There are two overloads that we need worry about.
+            // - object[] GetCustomAttributes(bool inherit)
+            // - object[] GetCustomAttributes(Type attributeType, bool inherit)
+
+            // Let's start with the first overload. It should return all the attributes without
+            // filtering on type.
             mockPropertyInfo.Setup(p => p.GetCustomAttributes(It.IsAny<bool>())).Returns(attributes);
 
+            // The second overload is a bit more complicated. This overload is also called internally
+            // by extension methods like propertyInfo.GetCustomAttribute<T> or GetCustomAttributes<T>()
+            // The overload should only return attributes of the specified type.
+            // We should ensure that we also return an array of the specific attribute type, even if it's empty.
+            // For example, if we call property.GetCustomAttributes<MyAttribute>(), we should return
+            // an instance of MyAttribute[], not an instance of object[] or Attribute[] that only contains MyAttribute values.
+            // This should be the case even if the array is empty. Which means we should mock any possible Attribute type.
+            // This is important because MyAttribute[] can always be causet to Attribute[], but Attribute[] cannot always be cast to MyAttribute[]
+            // Returning the wrong type will cause .NET to throw invalid cast exceptions.
+
+            // We first start with a general catch-call case that returns an empty array of the input attribute type
+            // for any attribute type the method is called with.
+            mockPropertyInfo.Setup(p => p.GetCustomAttributes(It.IsAny<Type>(), It.IsAny<bool>()))
+                .Returns((Type attributeType, bool inherit) => Activator.CreateInstance(attributeType.MakeArrayType(), [0]) as object[]);
+
+            // Then for attributes that have been passed, we create specific mocks to handle their types and return corresponding arrays.
+            foreach (var attribute in attributes)
+            {
+                // Let's say attribute is of type MyAttribute, the following is equivalent of doing:
+                // MyAttribute[] attributeArray = new MyAttribute[] { attribute };
+                var attributeType = attribute.GetType();
+                var attributeArrayType = attributeType.MakeArrayType();
+                var attributeArray = (object[])Activator.CreateInstance(attributeArrayType, [1]); // Create array of length 1
+                attributeArray[0] = attribute;
+                // This assumes that the property only has one instance of each property type. We don't handle the case where
+                // the multiple attributes of the same time are applied to the same property. If we ever have such a use case
```

Comment:
You meant `type` here
```suggestion
                // the multiple attributes of the same type are applied to the same property. If we ever have such a use case
```